<style  type="text/css">
cool {
  width: 100px;
  height: 100px;
  background-color: red;
  position: relative;
  -webkit-animation-name: example; /* Safari 4.0 - 8.0 */
  -webkit-animation-duration: 4s; /* Safari 4.0 - 8.0 */
  -webkit-animation-iteration-count: infinite; /* Safari 4.0 - 8.0 */
  animation-name: example;
  animation-duration: 4s;
  animation-iteration-count: infinite;
}

loud {
    font-size: 20pt;
    }

/* Safari 4.0 - 8.0 */
@-webkit-keyframes example {
  0%   {background-color:red; left:0px; top:0px;}
  25%  {background-color:yellow; left:200px; top:0px;}
  50%  {background-color:blue; left:200px; top:200px;}
  75%  {background-color:green; left:0px; top:200px;}
  100% {background-color:red; left:0px; top:0px;}
}

/* Standard syntax */
@keyframes example {
  0%   {background-color:red; left:0px; top:0px;}
  25%  {background-color:yellow; left:200px; top:0px;}
  50%  {background-color:blue; left:200px; top:200px;}
  75%  {background-color:green; left:0px; top:200px;}
  100% {background-color:red; left:0px; top:0px;}
}
</style>

<b><center>
<span style="font-size: 24pt; line-height: 1.2">
COMS W4111: Introduction to Databases<br>
Spring 2024, Sections 002/V02
</span>
</center></b>
<br>
<p>
<i><center>
<span style="font-size: 20pt; line-height: 1.2">
Homework 4<br>
</span>
</center></i>

# Introduction

- This notebook contains HW4. **Both Programming and Nonprogramming tracks should complete this homework.**
- You will submit **PDF and ZIP files** for this assignment. Gradescope will have two separate assignments for these.
- For the PDF:
    - The most reliable way to save as PDF is to go to your browser's menu bar and click `File -> Print`. Switch the orientation to landscape mode, and hit save.
    - **MAKE SURE ALL YOUR WORK (CODE AND SCREENSHOTS) IS VISIBLE ON THE PDF. YOU WILL NOT GET CREDIT IF ANYTHING IS CUT OFF.** Reach out for troubleshooting.
    - **MAKE SURE YOU DON'T SUBMIT A SINGLE PAGE PDF.** Your PDF should have multiple pages.
- For the ZIP:
    - Zip a folder containing this notebook and any screenshots.
    - You may delete any unnecessary files, such as caches.

# Setup

In [1]:
# %load_ext sql
%reload_ext sql
%sql mysql+pymysql://root:dbuserdbuser@localhost

In [4]:
import sys

!{sys.executable} -m pip install --upgrade pymongo
!{sys.executable} -m pip install --upgrade neo4j

  Obtaining dependency information for pymongo from https://files.pythonhosted.org/packages/2b/f0/bc30f0d9f7f1b3de1eb51ef8ed85378ff237f551c28bf4a23e71cf92aa20/pymongo-4.6.3-cp311-cp311-macosx_10_9_universal2.whl.metadata
  Obtaining dependency information for dnspython<3.0.0,>=1.16.0 from https://files.pythonhosted.org/packages/87/a1/8c5287991ddb8d3e4662f71356d9656d91ab3a36618c3dd11b280df0d255/dnspython-2.6.1-py3-none-any.whl.metadata
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 534.5/534.5 kB 12.3 MB/s eta 0:00:00a 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 307.7/307.7 kB 11.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 203.0/203.0 kB 6.8 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... done
  Created wheel for neo4j: filename=neo4j-5.19.0-py3-none-any.whl size=280741 sha256=bdf15f7e1605cebf27a952cf18ece8ecf21fcf735745b7

- If you get warnings below, try restarting your kernel

In [2]:
import neo4j
import pandas
import pymongo

# TODO: Fill in with your Mongo URL
mongo_url = "mongodb+srv://sh4350:dbuserdbuser@atlascluster.imktwuz.mongodb.net/?retryWrites=true&w=majority&appName=AtlasCluster"
mongo_client = pymongo.MongoClient(mongo_url)

# TODO: Fill in with your Neo4j credentials
neo4j_url = "neo4j+ssc://2f2ab41e.databases.neo4j.io"
neo4j_password = "yec69zxPbB5STpSZQwoNUVOfK1cXzajpek64qQiXBR4"
# username is always "neo4j"
graph = neo4j.GraphDatabase.driver(neo4j_url, auth=("neo4j", neo4j_password))
graph.verify_connectivity()

<hr style="height:2px;">

# Written Questions

- As usual, do not bloviate

## W1

Explain the following concepts:

1. Clustering index
2. Nonclustering index
3. Sparse index
4. Dense index


Clustering Index: Organizes the actual data records in the database based on the key values of the index, efficient for range queries, but insertion and deletion can be costly because they may require reorganizing the data.

Nonclustering Index: Contains index entries that point to the actual data records, which are stored separately, multiple nonclustering indexes can exist on a table, but may lead to slower range queries compared to clustering indexes.

Sparse Index: Only index entries for some of the records based on a certain interval, requires less space, can be slower for lookups since it might need to scan through more data compared to a dense index.

Dense Index: Contains an index entry for every single record in the database, faster lookups for any given record, but takes up more space than a sparse index.

## W2

Explain why nonclustering indexes must be dense.

Nonclustering indexes must be dense to ensure that every record in the table can be efficiently located. Since the data records are stored separately from the index and not in sorted order, a dense index—with an entry for every record—guarantees that any search or lookup can directly access the corresponding record without missing any, ensuring comprehensive and efficient data retrieval.

## W3

Suppose that, in a table containing information about Columbia classes, the columns `class_code`, `semester`, and `year` are queried frequently **individually**. Would putting a composite index on `(class_code, semester, year)` be a good idea? Why or why not?


No, putting a composite index on (class_code, semester, year) would not be the best idea if the columns are queried frequently individually. A composite index is most effective when queries involve all the columns in the index or the leftmost subset. Queries on individual columns won't benefit optimally from a composite index structured this way, because the index's effectiveness diminishes when the query does not start with the first column in the composite index sequence. It would be more beneficial to create separate indexes on each column to optimize their individual query performance.

## W4

Explain the following concepts:

1. Hash index
2. B+ tree index

Hash Index: buckets store entries with pointers to records, ideal for quick lookups of specific items, Best for exact match searches, but not suited for range queries or maintaining order.
B+ Tree Index: Tree-based structure with nodes; optimizes reading/writing large blocks of sorted data, supports a wide range of queries, including point, range, and sequential access, efficient for operations involving large data sets, maintaining order, and facilitating quick inserts and deletes.

## W5

Give one advantage and one disadvantage of hash indexes compared to B+ tree indexes.


Advantage of Hash Indexes: Faster data retrieval for exact-match queries due to direct computation.

Disadvantage of Hash Indexes: Inefficient for range queries, as they do not maintain any sort of order among keys.

## W6

Explain the role of the buffer in a DBMS. Why doesn't the DBMS simply load the entire database in its buffer?


The buffer in a DBMS temporarily holds data from the disk to reduce the number of disk I/O operations, speeding up read/write operations since accessing data in memory is much faster than disk access. A DBMS doesn't load the entire database into its buffer due to limitations in physical memory size and the inefficiency of loading data that might not be needed immediately or frequently, which would waste valuable memory resources and potentially slow down the system.

## W7

Explain the following concepts as they relate to buffer replacement policies:

1. Clairvoyant algorithm
2. Least recently used strategy
3. Most recently used strategy
4. Clock algorithm


Clairvoyant Algorithm: An ideal, theoretical approach that always evicts the page that will be used farthest in the future. It requires future knowledge of requests, serving as a benchmark for other algorithms' performance.

Least Recently Used (LRU) Strategy: Evicts the page that has been unused for the longest time, based on the assumption that pages used recently will likely be used again soon.

Most Recently Used (MRU) Strategy: Opposite of LRU, it evicts the most recently used page, based on the premise that the most recently accessed page will be the least likely to be accessed again in the near future.

Clock Algorithm: A practical approximation of LRU that organizes pages in a circular manner (like a clock) and uses a "hand" that points to the oldest page. Pages are given a second chance before eviction if they are marked as recently used, efficiently managing memory with lower overhead.

## W8

NoSQL databases have become increasingly popular for applications. List 3 benefits of using NoSQL databases over SQL ones.


Scalability: NoSQL databases are designed to scale out using distributed architectures, making them well-suited for handling large volumes of data and traffic.

Flexibility: They often support schema-less data models, allowing for the storage of unstructured or semi-structured data without a predefined schema, which is beneficial for rapidly evolving applications.

Performance: NoSQL databases can provide faster data retrieval and handling for specific types of queries and data models, particularly when dealing with large volumes of data that don't fit well into relational models.

## W9

Explain the concept between impedance mismatch and how it relates to SQL vs. NoSQL databases.


Impedance mismatch refers to the conflict between the object-oriented model used in application programming and the relational model used in SQL databases. This discrepancy arises because objects in code do not neatly map onto the tabular structure of relational databases, requiring additional effort to translate between the two models (e.g., using Object-Relational Mapping (ORM) tools). NoSQL databases, with their more flexible data models (such as document, key-value, graph), often align more closely with the object-oriented approach, reducing or eliminating the impedance mismatch and simplifying the interaction between application code and the database.

## W10

The relationship between students and courses is many-to-many. Due to its emphasis on atomicity, modeling this relationship in a relational database would require an associative entity. Explain how this relationship could be modeled in

1. A document database, such as MongoDB
2. A graph database, such as Neo4j


In MongoDB (Document Database): This many-to-many relationship can be modeled using embedded documents or references. For instance, each Student document could contain an array of references (IDs) to Course documents they are enrolled in, or vice versa. 

In Neo4j (Graph Database): The relationship is naturally modeled by creating Student and Course nodes, with edges (relationships) connecting them. Each edge can represent an enrollment, directly mapping the many-to-many relationship. Additional properties can be added to these edges (such as enrollment date) to capture more details about the relationship.

<hr style="height:2px;">

# MongoDB

- The cell below creates a database `w4111`, then a collection `episodes` inside `w4111`. It then inserts GoT episode data into the collection.

In [7]:
import json

with open("episodes.json") as f:
    data = json.load(f)

episodes = mongo_client["w4111"]["episodes"]    
episodes.drop()
episodes.insert_many(data)
print("Successfully inserted episode data")

Successfully inserted episode data


## M1

- Write and execute a query that shows episodes and the number of scenes they contain
- Your aggregation should have the following attributes:
    - `episodeTitle`
    - `seasonNum`
    - `episodeNum`
    - `numScenes`, which is the length of the episode's `scenes` array
- Order your output on `numScenes` descending, and only keep episodes with more than 100 scenes

In [16]:
res = episodes.aggregate([
    {
        "$project": {
            "episodeTitle": 1, 
            "seasonNum": 1,    
            "episodeNum": 1,   
            "numScenes": {"$size": "$scenes"}  
        }
    },
    {
        "$match": {
            "numScenes": {"$gt": 100}  
        }
    },
    {
        "$sort": {
            "numScenes": -1  
        }
    },
    {
    '$project': {
        '_id': 0,
        'episodeTitle': '$episodeTitle',
        'seasonNum': '$seasonNum',
        'episodeNum': '$episodeNum',
        'numScenes': '$numScenes'
    }
}


])

pandas.DataFrame(list(res))

episodeTitle  seasonNum  episodeNum  numScenes
0           The Long Night          8           3        292
1                The Bells          8           5        220
2               Blackwater          2           9        133
3   The Last of the Starks          8           4        113
4  The Dragon and the Wolf          7           7        104

## M2

- Write and execute a query that shows the first three episodes for each season
- Your aggregation should have the following attributes:
    - `seasonNum`
    - `firstThreeEpisodes`, which is an array that contains the titles of the first, second, and third episodes (in that order) of the season
- Order your output on `seasonNum` ascending
    - It's okay if the `firstThreeEpisodes` column is a bit truncated by the dataframe

In [12]:
res = episodes.aggregate([
    {
        "$sort": {
            "seasonNum": 1,  # Sort by season number ascending
            "episodeNum": 1  # Sort by episode number ascending within each season
        }
    },
    {
        "$match": {
            "episodeNum": {"$lte": 3}  # Include only the first three episodes of each season
        }
    },
    {
        "$group": {
            "_id": "$seasonNum",  # Group by season number
            "firstThreeEpisodes": {"$push": "$episodeTitle"}  # Collect episode titles in order
        }
    },
    {
        "$project": {
            "seasonNum": "$_id",  # Include the season number
            "firstThreeEpisodes": 1  # Include the array of episode titles
        }
    },
    {
        "$sort": {
            "seasonNum": 1  # Sort by season number ascending in the final output
        }
    }
])


pandas.DataFrame(list(res))

_id                                 firstThreeEpisodes  seasonNum
0    1       [Winter Is Coming, The Kingsroad, Lord Snow]          1
1    2  [The North Remembers, The Night Lands, What Is...          2
2    3  [Valar Dohaeris, Dark Wings, Dark Words, Walk ...          3
3    4  [Two Swords, The Lion and the Rose, Breaker of...          4
4    5  [The Wars to Come, The House of Black and Whit...          5
5    6                 [The Red Woman, Home, Oathbreaker]          6
6    7      [Dragonstone, Stormborn, The Queen's Justice]          7
7    8  [Winterfell, A Knight of the Seven Kingdoms, T...          8

## M3

- Write and execute a query that shows statistics about each season
- Your aggregation should have the following attributes:
    - `seasonNum`
    - `numEpisodes`, which is the number of episodes in the season
    - `startDate`, which is the earliest air date associated with an episode in the season
    - `endDate`, which is the latest air date associated with an episode in the season
    - `shortestEpisodeLength`
    - `longestEpisodeLength`
        - The length of an episode is the greatest `sceneEnd` value in the episode's `scenes` array
- Order your output on `seasonNum` ascending

In [13]:
res = episodes.aggregate([
    {
        "$unwind": "$scenes"  # Unwind the scenes array to process each scene individually
    },
    {
        "$addFields": {
            "sceneEndSec": {
                "$sum": [
                    {"$multiply": [{"$toInt": {"$arrayElemAt": [{"$split": ["$scenes.sceneEnd", ":"]}, 0]}}, 3600]},
                    {"$multiply": [{"$toInt": {"$arrayElemAt": [{"$split": ["$scenes.sceneEnd", ":"]}, 1]}}, 60]},
                    {"$toInt": {"$arrayElemAt": [{"$split": ["$scenes.sceneEnd", ":"]}, 2]}}
                ]
            }
        }
    },
    {
        "$group": {
            "_id": {
                "seasonNum": "$seasonNum",
                "episodeNum": "$episodeNum",
                "episodeAirDate": "$episodeAirDate"
            },
            "maxSceneEnd": {"$max": "$sceneEndSec"}
        }
    },
    {
        "$group": {
            "_id": "$_id.seasonNum",
            "numEpisodes": {"$sum": 1},
            "startDate": {"$min": "$_id.episodeAirDate"},
            "endDate": {"$max": "$_id.episodeAirDate"},
            "shortestEpisodeLength": {"$min": "$maxSceneEnd"},
            "longestEpisodeLength": {"$max": "$maxSceneEnd"}
        }
    },
    {
        "$project": {
            "seasonNum": "$_id",
            "numEpisodes": 1,
            "startDate": 1,
            "endDate": 1,
            "shortestEpisodeLength": 1,
            "longestEpisodeLength": 1
        }
    },
    {
        "$sort": {
            "seasonNum": 1
        }
    }
])

pandas.DataFrame(list(res))

_id  numEpisodes   startDate     endDate  shortestEpisodeLength  \
0    1           10  2011-04-17  2011-06-19                   3090   
1    2           10  2012-04-01  2012-06-03                   2958   
2    3           10  2013-03-31  2013-06-09                   2965   
3    4           10  2014-04-06  2014-06-15                   2959   
4    5           10  2015-03-29  2015-06-14                   3032   
5    6           10  2016-04-24  2016-06-26                   3112   
6    7            7  2017-07-16  2017-08-27                   3005   
7    8            6  2019-04-14  2019-05-19                   3269   

   longestEpisodeLength  seasonNum  
0                  3657          1  
1                  3724          2  
2                  3680          3  
3                  3889          4  
4                  3721          5  
5                  4214          6  
6                  4870          7  
7                  4897          8

## M4

- Write and execute a query that shows sublocations and the scenes they appear in
- Your aggregation should have the following attributes:
    - `subLocation`
    - `totalScenes`, which is the number of scenes that are set in the sublocation
    - `firstSeasonNum`
    - `firstEpisodeNum`
        - `(firstSeasonNum, firstEpisodeNum)` identifies the first episode that the sublocation appears in
    - `lastSeasonNum`
    - `lastEpisodeNum`
        - `(lastSeasonNum, lastEpisodeNum)` identifies the last episode that the sublocation appears in
- Order your output on `totalScenes` descending, and only keep the sublocations with more than 50 scenes

In [14]:
res = episodes.aggregate([
    {
        "$unwind": "$scenes"  # Flatten the scenes array
    },
    {
        "$group": {
            "_id": "$scenes.subLocation",  # Group by sublocation
            "totalScenes": {"$sum": 1},  # Count the number of scenes
            "firstSeasonNum": {"$min": "$seasonNum"},  # Minimum season number for the first appearance
            "firstEpisodeNum": {"$first": "$episodeNum"},  # Episode number of the first appearance in the group
            "lastSeasonNum": {"$max": "$seasonNum"},  # Maximum season number for the last appearance
            "lastEpisodeNum": {"$last": "$episodeNum"}  # Episode number of the last appearance in the group
        }
    },
    {
        "$match": {
            "totalScenes": {"$gt": 50}  # Filter sublocations with more than 50 scenes
        }
    },
    {
        "$sort": {
            "totalScenes": -1  # Sort by the total number of scenes in descending order
        }
    },
    {
        "$project": {
            "subLocation": "$_id",
            "totalScenes": 1,
            "firstSeasonNum": 1,
            "firstEpisodeNum": 1,
            "lastSeasonNum": 1,
            "lastEpisodeNum": 1
        }
    }
])

pandas.DataFrame(list(res))

_id  totalScenes  firstSeasonNum  firstEpisodeNum  \
0       King's Landing         1094               1                1   
1           Winterfell          734               1                1   
2                 None          494               1                1   
3         Castle Black          267               1                1   
4          Dragonstone          142               2                1   
5   The Haunted Forest           77               1                1   
6   Outside Winterfell           69               1                1   
7       Craster's Keep           66               2                1   
8             The Wall           60               2               10   
9            The Twins           57               1                9   
10     Blackwater Rush           56               7                4   
11      Blackwater Bay           53               2                8   

    lastSeasonNum  lastEpisodeNum         subLocation  
0               8               6      King's Landing  
1               8               6          Winterfell  
2               8               6                None  
3               8               6        Castle Black  
4               8               5         Dragonstone  
5               8               6  The Haunted Forest  
6               8               4  Outside Winterfell  
7               4               5      Craster's Keep  
8               8               6            The Wall  
9               7               1           The Twins  
10              7               5     Blackwater Rush  
11              8               6      Blackwater Bay

<hr style="height:2px;">

# Neo4j

- The cell below creates nodes and relationships that model movies and the people involved in them

In [25]:
with open("movies.txt") as f:
    queries = str(f.read())

graph.execute_query("match (p:Person), (m:Movie) detach delete p, m")
graph.execute_query(queries)
print("Successfully inserted movie data")

Successfully inserted movie data


## N1

- Write and execute a cypher that shows actors and the number of movies they appear in
    - You should focus only on the `ACTED_IN` relationship, no other relationship
- Your output should have the following attributes:
    - `name`, which is the name of the actor
    - `num_movies`
- Order your output on `num_movies` descending, and only keep actors who have acted in 4 or more movies

In [32]:
res = graph.execute_query("""
    MATCH (a:Person)-[rel:ACTED_IN]->(m:Movie)
    WITH a.name AS actor, COUNT(DISTINCT m) AS num_movies
    WHERE num_movies >= 4
    RETURN actor, num_movies
    ORDER BY num_movies DESC
""")

pandas.DataFrame([dict(r) for r in res.records])

actor  num_movies
0         Tom Hanks          12
1      Keanu Reeves           7
2      Hugo Weaving           5
3    Jack Nicholson           5
4          Meg Ryan           5
5  Cuba Gooding Jr.           4

## N2

- Write and execute a cypher that shows people and movies they either acted in or directed
- Your output should have the following attributes:
    - `name`, which is the name of the person
    - `directed_movies`, which is an array of titles of movies that the person directed
    - `acted_in_movies`, which is an array of titles of movies that the person acted in
- Order your output on `name` ascending, and only keep people that have directed at least one movie **and** acted in at least one movie (i.e., there should be no empty arrays. Arrays with one element are fine.)

In [33]:
res = graph.execute_query("""
    MATCH (p:Person)-[:ACTED_IN]->(a:Movie)
    WITH p, COLLECT(a.title) AS acted_in_movies
    MATCH (p)-[:DIRECTED]->(d:Movie)
    WITH p.name AS name, acted_in_movies, COLLECT(d.title) AS directed_movies
    WHERE SIZE(acted_in_movies) > 0 AND SIZE(directed_movies) > 0
    RETURN name, directed_movies, acted_in_movies
    ORDER BY name ASC

""")

pandas.DataFrame([dict(r) for r in res.records])

name                   directed_movies  \
0  Clint Eastwood                      [Unforgiven]   
1    Danny DeVito                           [Hoffa]   
2  James Marshall  [V for Vendetta, Ninja Assassin]   
3       Tom Hanks               [That Thing You Do]   
4   Werner Herzog                      [RescueDawn]   

                                     acted_in_movies  
0                                       [Unforgiven]  
1           [Hoffa, One Flew Over the Cuckoo's Nest]  
2                                   [A Few Good Men]  
3  [You've Got Mail, Sleepless in Seattle, Joe Ve...  
4                             [What Dreams May Come]

## N3

- Write and execute a cypher that shows people and movies they both acted in and directed
- Your output should have the following attributes:
    - `name`, which is the name of the person
    - `acted_in_and_directed_movies`, which is an array of titles of movies that the person both acted in and directed
- Order your output on `name` ascending, and only keep people that have acted in at least one movie that they directed (i.e., there should be no empty arrays. Arrays with one element are fine.)

In [39]:
res = graph.execute_query("""
    MATCH (p:Person)-[:ACTED_IN]->(m:Movie)
    WHERE (p)-[:DIRECTED]->(m)
    WITH p, COLLECT(m.title) AS acted_in_and_directed_movies
    WHERE SIZE(acted_in_and_directed_movies) > 0
    RETURN p.name AS name, acted_in_and_directed_movies
    ORDER BY name ASC

""")

pandas.DataFrame([dict(r) for r in res.records])

name acted_in_and_directed_movies
0  Clint Eastwood                 [Unforgiven]
1    Danny DeVito                      [Hoffa]
2       Tom Hanks          [That Thing You Do]

## N4

- Write and execute a cypher that shows pairs of people and how closely connected they are
- Your output should have the following attributes:
    - `person_1_name`, which is the name of the first person in the pair
    - `person_2_name`, which is the name of the second person in the pair
    - `num_people_between`, which is the number of people (including the pair itself) separating the pair. You should use the `shortestPath` function to compute this.
- To prevent duplicates in your output, you should only keep rows where `person_1_name < person_2_name`
- Order your output on `(person_1_name, person_2_name)`, and only keep rows where `num_people_between > 5`


- As an example, you should get the following row in your output:

| person_1_name | person_2_name | num_people_between |
|---------------|---------------|--------------------|
| Billy Crystal | Paul Blythe   | 6                  |

- The shortest path between Billy Crystal and Paul Blythe is shown below
    - `num_people_between` is 6 because there are 6 nodes marked as `Person` (including Billy's and Paul's nodes)

<img src="./N4-example.png">

In [44]:
res = graph.execute_query("""
    MATCH path=shortestPath((p1:Person)-[*]-(p2:Person))
    WHERE p1.name < p2.name
    WITH p1, p2, path
    WHERE LENGTH(path) >= 5  
    RETURN p1.name AS person_1_name, p2.name AS person_2_name, LENGTH(path) + 1 AS num_people_between
    ORDER BY person_1_name, person_2_name

""")

# pandas.DataFrame([dict(r) for r in res.records])

# Assume `res` is your result from the Neo4j query execution
data = [dict(r) for r in res.records]  # Convert records to a list of dictionaries
df = pandas.DataFrame(data)  # Create a DataFrame from the list

# Filter to find the specific pair
filtered_df = df[(df['person_1_name'] == 'Billy Crystal') & (df['person_2_name'] == 'Paul Blythe')]

# Display the filtered DataFrame
print(filtered_df)

     person_1_name person_2_name  num_people_between
752  Billy Crystal   Paul Blythe                   9


<hr style="height:2px;">

# SQL To NoSQL

- You will move relational data to document and graph databases
    - **You will do your modeling in Python. You shouldn't be writing any SQL.**
- You will be using the `classicmodels` database for this section. You may want to drop the database and re-run the SQL script (included in the directory) to ensure you have the right data.
    - You will be modeling customers and the products they ordered

In [45]:
# %%sql 

# DROP DATABASE IF EXISTS classicmodels;
# CREATE DATABASE classicmodels;
# USE classicmodels;



 * mysql+pymysql://root:***@localhost
8 rows affected.
1 rows affected.
0 rows affected.


[]

## MongoDB: Customers

- For the document database, you will create two collections: `customers` and `products`
- `customers` will contain customer information as well as all the orders they've placed
- You will use `customer_orders_all_df` to create your `customers` collection

In [3]:
%%sql

customer_orders_all <<

select
    c.customerNumber, c.customerName, c.country, o.orderNumber,
    o.orderDate, od.productCode, od.quantityOrdered, od.priceEach
from classicmodels.orders o
    join classicmodels.customers c using (customerNumber)
    join classicmodels.orderdetails od using (orderNumber);

 * mysql+pymysql://root:***@localhost
2996 rows affected.
Returning data to local variable customer_orders_all


In [4]:
customer_orders_all_df = customer_orders_all.DataFrame()
customer_orders_all_df.head(10)

customerNumber                  customerName  country  orderNumber  \
0             363  Online Diecast Creations Co.      USA        10100   
1             363  Online Diecast Creations Co.      USA        10100   
2             363  Online Diecast Creations Co.      USA        10100   
3             363  Online Diecast Creations Co.      USA        10100   
4             128          Blauer See Auto, Co.  Germany        10101   
5             128          Blauer See Auto, Co.  Germany        10101   
6             128          Blauer See Auto, Co.  Germany        10101   
7             128          Blauer See Auto, Co.  Germany        10101   
8             181               Vitachrome Inc.      USA        10102   
9             181               Vitachrome Inc.      USA        10102   

    orderDate productCode  quantityOrdered priceEach  
0  2003-01-06    S18_1749               30    136.00  
1  2003-01-06    S18_2248               50     55.09  
2  2003-01-06    S18_4409               22     75.46  
3  2003-01-06    S24_3969               49     35.29  
4  2003-01-09    S18_2325               25    108.06  
5  2003-01-09    S18_2795               26    167.06  
6  2003-01-09    S24_1937               45     32.53  
7  2003-01-09    S24_2022               46     44.35  
8  2003-01-10    S18_1342               39     95.55  
9  2003-01-10    S18_1367               41     43.13

- Below is an example of how a customer and their orders are stored in MySQL, and how the document should look like in MongoDB
- The document should have the following attributes:
    - `customerNumber`
    - `customerName`
    - `country`
    - `orders`, which is a list of objects. Each object represents one order
        - `orderNumber`
        - `orderDate`
        - `orderContents`, which is a list of objects. Each object represents one product in the order
            - `productCode`
            - `quantityOrdered`
            - `priceEach`
            

**MySQL relation:**

|    |   customerNumber | customerName      | country   |   orderNumber | orderDate   | productCode   |   quantityOrdered |   priceEach |
|---:|-----------------:|:------------------|:----------|--------------:|:------------|:--------------|------------------:|------------:|
|  0 |              103 | Atelier graphique | France    |         10123 | 2003-05-20  | S18_1589      |                26 |      120.71 |
|  1 |              103 | Atelier graphique | France    |         10123 | 2003-05-20  | S18_2870      |                46 |      114.84 |
|  2 |              103 | Atelier graphique | France    |         10123 | 2003-05-20  | S18_3685      |                34 |      117.26 |
|  3 |              103 | Atelier graphique | France    |         10123 | 2003-05-20  | S24_1628      |                50 |       43.27 |
|  4 |              103 | Atelier graphique | France    |         10298 | 2004-09-27  | S10_2016      |                39 |      105.86 |
|  5 |              103 | Atelier graphique | France    |         10298 | 2004-09-27  | S18_2625      |                32 |       60.57 |
|  6 |              103 | Atelier graphique | France    |         10345 | 2004-11-25  | S24_2022      |                43 |       38.98 |



**MongoDB document:**

```
{
    customerNumber: 103
    customerName: "Atelier graphique",
    country: "France",
    orders: [
        {
            orderNumber: 10123,
            orderDate: "2003-05-20",
            orderContents: [
                {
                    productCode: "S18_1589",
                    quantityOrdered: 26,
                    priceEach: "120.71"
                },
                {
                    productCode: "S18_2870",
                    quantityOrdered: 46,
                    priceEach: "114.84"
                },
                {
                    productCode: "S18_3685",
                    quantityOrdered: 34,
                    priceEach: "117.26"
                },
                {
                    productCode: "S24_1628",
                    quantityOrdered: 50,
                    priceEach: "43.27"
                }
            ]
        },
        {
            orderNumber: 10298,
            orderDate: "2004-09-27",
            orderContents: [
                {
                    productCode: "S10_2016",
                    quantityOrdered: 39,
                    priceEach: "105.86"
                },
                {
                    productCode: "S18_2625",
                    quantityOrdered: 32,
                    priceEach: "60.57"
                }
            ]
        },
        {
            orderNumber: 10345,
            orderDate: "2004-11-25",
            orderContents: [
                {
                    productCode: "S24_2022",
                    quantityOrdered: 43,
                    priceEach: "38.98"
                }
            ]
        }
    ]
}
```

In [5]:
# TODO: Create a list of dicts. Each dict represents one customer.

"""
Tips:

    To iterate through dataframe:
    
        for _, r in customer_orders_all_df.iterrows():
            r = dict(r)
            Access fields like r['customerName'], r['country'], ...
            
    
    The orderDate and priceEach fields are stored as datetime.date and Decimal
    objects in the dataframe. These types are not compatible with the pymongo API.
    You can convert them to strings by calling str(r['orderDate']) and str(r['priceEach']). 
    Alternatively, you can look into the datetime.datetime and bson.decimal128.Decimal128 
    objects, which are supported by pymongo.

"""
customers = []

df = customer_orders_all_df

for customer_number, customer_group in df.groupby('customerNumber'):
    customer_info = customer_group.iloc[0]  # Get common customer information
    customer_dict = {
        'customerNumber': customer_number,
        'customerName': customer_info['customerName'],
        'country': customer_info['country'],
        'orders': []
    }
    
    for order_number, order_group in customer_group.groupby('orderNumber'):
        order_dict = {
            'orderNumber': order_number,
            'orderDate': str(order_group.iloc[0]['orderDate']),
            'orderContents': []
        }
        
        # Iterate over each product in the order
        for _, product_row in order_group.iterrows():
            product_dict = {
                'productCode': product_row['productCode'],
                'quantityOrdered': product_row['quantityOrdered'],
                'priceEach': str(product_row['priceEach'])  # Convert Decimal to string if necessary
            }
            order_dict['orderContents'].append(product_dict)
        
        customer_dict['orders'].append(order_dict)
    
    customers.append(customer_dict)


In [7]:
customers

[{'customerNumber': 103,
  'customerName': 'Atelier graphique',
  'country': 'France',
  'orders': [{'orderNumber': 10123,
    'orderDate': '2003-05-20',
    'orderContents': [{'productCode': 'S18_1589',
      'quantityOrdered': 26,
      'priceEach': '120.71'},
     {'productCode': 'S18_2870', 'quantityOrdered': 46, 'priceEach': '114.84'},
     {'productCode': 'S18_3685', 'quantityOrdered': 34, 'priceEach': '117.26'},
     {'productCode': 'S24_1628',
      'quantityOrdered': 50,
      'priceEach': '43.27'}]},
   {'orderNumber': 10298,
    'orderDate': '2004-09-27',
    'orderContents': [{'productCode': 'S10_2016',
      'quantityOrdered': 39,
      'priceEach': '105.86'},
     {'productCode': 'S18_2625',
      'quantityOrdered': 32,
      'priceEach': '60.57'}]},
   {'orderNumber': 10345,
    'orderDate': '2004-11-25',
    'orderContents': [{'productCode': 'S24_2022',
      'quantityOrdered': 43,
      'priceEach': '38.98'}]}],
  '_id': ObjectId('662191dbbfa91778b86e3e13')},
 {'custom

In [6]:
def insert_customers(d):
    mongo_client['w4111']['customers'].drop()
    mongo_client['w4111']['customers'].insert_many(d)
    

# TODO: Put the name of your list of dicts below
insert_customers(customers)
print("Successfully inserted customer data")

Successfully inserted customer data


## MongoDB: Products

- To create the `products` collection, you will use `products_all_df`
- A document in `products` simply contains product information, as shown below

```
{
    productCode: "S10_1678",
    productName: "1969 Harley Davidson Ultimate Chopper",
    productVendor: "Min Lin Diecast"
}
```

In [8]:
%%sql

products_all <<

select productCode, productName, productVendor
from classicmodels.products;

 * mysql+pymysql://root:***@localhost
110 rows affected.
Returning data to local variable products_all


In [9]:
products_all_df = products_all.DataFrame()
products_all_df.head(10)

productCode                            productName  \
0    S10_1678  1969 Harley Davidson Ultimate Chopper   
1    S10_1949               1952 Alpine Renault 1300   
2    S10_2016                  1996 Moto Guzzi 1100i   
3    S10_4698   2003 Harley-Davidson Eagle Drag Bike   
4    S10_4757                    1972 Alfa Romeo GTA   
5    S10_4962                 1962 LanciaA Delta 16V   
6    S12_1099                      1968 Ford Mustang   
7    S12_1108                      2001 Ferrari Enzo   
8    S12_1666                         1958 Setra Bus   
9    S12_2823                       2002 Suzuki XREO   

               productVendor  
0            Min Lin Diecast  
1    Classic Metal Creations  
2   Highway 66 Mini Classics  
3          Red Start Diecast  
4    Motor City Art Classics  
5        Second Gear Diecast  
6      Autoart Studio Design  
7        Second Gear Diecast  
8  Welly Diecast Productions  
9       Unimax Art Galleries

In [10]:
# TODO: Create a list of dicts. Each dict represents one product.

"""
Tips:

    To iterate through dataframe:
    
        for _, r in products_all_df.iterrows():
            r = dict(r)
            Access fields like r['productName'], r['productVendor'], ...

"""

products = []

for _, r in products_all_df.iterrows():
    r = dict(r)
    product_dict = {}
    for attribute in ['productCode', 'productName', 'productVendor']:
        product_dict[attribute] = r[attribute]

    products.append(product_dict)

products 

[{'productCode': 'S10_1678',
  'productName': '1969 Harley Davidson Ultimate Chopper',
  'productVendor': 'Min Lin Diecast'},
 {'productCode': 'S10_1949',
  'productName': '1952 Alpine Renault 1300',
  'productVendor': 'Classic Metal Creations'},
 {'productCode': 'S10_2016',
  'productName': '1996 Moto Guzzi 1100i',
  'productVendor': 'Highway 66 Mini Classics'},
 {'productCode': 'S10_4698',
  'productName': '2003 Harley-Davidson Eagle Drag Bike',
  'productVendor': 'Red Start Diecast'},
 {'productCode': 'S10_4757',
  'productName': '1972 Alfa Romeo GTA',
  'productVendor': 'Motor City Art Classics'},
 {'productCode': 'S10_4962',
  'productName': '1962 LanciaA Delta 16V',
  'productVendor': 'Second Gear Diecast'},
 {'productCode': 'S12_1099',
  'productName': '1968 Ford Mustang',
  'productVendor': 'Autoart Studio Design'},
 {'productCode': 'S12_1108',
  'productName': '2001 Ferrari Enzo',
  'productVendor': 'Second Gear Diecast'},
 {'productCode': 'S12_1666',
  'productName': '1958 Se

In [11]:
def insert_products(d):
    mongo_client['w4111']['products'].drop()
    mongo_client['w4111']['products'].insert_many(d)
    

# TODO: Put the name of your list of dicts below
insert_products(products)
print("Successfully inserted product data")

Successfully inserted product data


## MongoDB: Testing

- Run through the following cells
- **Make sure the outputs are completely visible. You shouldn't need to scroll to see the entire output.**
    - You may need to click on the blank section immediately to the left of your output to toggle between scrolling and unscrolling

In [12]:
import json

def prepr(doc):
    try:
        del doc['_id']
    except KeyError:
        pass
    
    def convert_str(d):
        if isinstance(d, dict):
            for k, v in d.items():
                d[k] = convert_str(v)
            return d
        elif isinstance(d, list):
            for i, v in enumerate(d):
                d[i] = convert_str(v)
            return d
        else:
            return str(d)
                
    convert_str(doc)
    return json.dumps(doc, indent=2)

In [13]:
res = mongo_client['w4111']['customers'].aggregate([
    {
        '$match': {
            'customerNumber': 219
        }
    }
])

print(prepr(list(res)[0]))

{
  "customerNumber": "219",
  "customerName": "Boards & Toys Co.",
  "country": "USA",
  "orders": [
    {
      "orderNumber": "10154",
      "orderDate": "2003-10-02",
      "orderContents": [
        {
          "productCode": "S24_3151",
          "quantityOrdered": "31",
          "priceEach": "75.23"
        },
        {
          "productCode": "S700_2610",
          "quantityOrdered": "36",
          "priceEach": "59.27"
        }
      ]
    },
    {
      "orderNumber": "10376",
      "orderDate": "2005-02-08",
      "orderContents": [
        {
          "productCode": "S12_3380",
          "quantityOrdered": "35",
          "priceEach": "98.65"
        }
      ]
    }
  ]
}


In [14]:
res = mongo_client['w4111']['customers'].aggregate([
    {
        '$match': {
            'customerNumber': 103
        }
    }
])

print(prepr(list(res)[0]))

{
  "customerNumber": "103",
  "customerName": "Atelier graphique",
  "country": "France",
  "orders": [
    {
      "orderNumber": "10123",
      "orderDate": "2003-05-20",
      "orderContents": [
        {
          "productCode": "S18_1589",
          "quantityOrdered": "26",
          "priceEach": "120.71"
        },
        {
          "productCode": "S18_2870",
          "quantityOrdered": "46",
          "priceEach": "114.84"
        },
        {
          "productCode": "S18_3685",
          "quantityOrdered": "34",
          "priceEach": "117.26"
        },
        {
          "productCode": "S24_1628",
          "quantityOrdered": "50",
          "priceEach": "43.27"
        }
      ]
    },
    {
      "orderNumber": "10298",
      "orderDate": "2004-09-27",
      "orderContents": [
        {
          "productCode": "S10_2016",
          "quantityOrdered": "39",
          "priceEach": "105.86"
        },
        {
          "productCode": "S18_2625",
          "quantityOrd

In [15]:
res = mongo_client['w4111']['products'].aggregate([
    {
        '$match': {
            'productCode': 'S18_1889'
        }
    }
])

print(prepr(list(res)[0]))

{
  "productCode": "S18_1889",
  "productName": "1948 Porsche 356-A Roadster",
  "productVendor": "Gearbox Collectibles"
}


## Neo4j: All Data

- For the graph database, you will have two types of nodes: `Customer` and `Product`
    - **Make sure to use these exact names**
- An order is represented as a relationship from the `Customer` node to the `Product` node. The type of the relationship should be `ORDERED`.


- You will use `customer_orders_limit_df` to create your graph

In [16]:
%%sql

customer_orders_limit <<

select
    c.customerNumber, c.customerName, c.country, o.orderNumber,
    o.orderDate, od.productCode, p.productName, p.productVendor,
    od.quantityOrdered, od.priceEach
from classicmodels.orders o
    join classicmodels.customers c using (customerNumber)
    join classicmodels.orderdetails od using (orderNumber)
    join classicmodels.products p using (productCode)
where od.quantityOrdered > 49;

 * mysql+pymysql://root:***@localhost
139 rows affected.
Returning data to local variable customer_orders_limit


In [17]:
customer_orders_limit_df = customer_orders_limit.DataFrame()
customer_orders_limit_df.head(10)

customerNumber                  customerName    country  orderNumber  \
0             363  Online Diecast Creations Co.        USA        10100   
1             145      Danish Wholesale Imports    Denmark        10105   
2             145      Danish Wholesale Imports    Denmark        10105   
3             278                 Rovelli Gifts      Italy        10106   
4             124  Mini Gifts Distributors Ltd.        USA        10113   
5             148       Dragon Souveniers, Ltd.  Singapore        10117   
6             353            Reims Collectables     France        10121   
7             103             Atelier graphique     France        10123   
8             458    Corrida Auto Replicas, Ltd      Spain        10126   
9             324      Stylish Desk Decors, Co.         UK        10129   

    orderDate productCode                productName  \
0  2003-01-06    S18_2248         1911 Ford Town Car   
1  2003-02-11    S10_4757        1972 Alfa Romeo GTA   
2  2003-02-11    S24_3816   1940 Ford Delivery Sedan   
3  2003-02-17    S24_3949   Corsair F4U ( Bird Cage)   
4  2003-03-26    S18_4668    1939 Cadillac Limousine   
5  2003-04-16    S72_3212                 Pont Yacht   
6  2003-05-07    S12_2823           2002 Suzuki XREO   
7  2003-05-20    S24_1628  1966 Shelby Cobra 427 S/C   
8  2003-05-28    S18_4600           1940s Ford truck   
9  2003-06-12    S24_3816   1940 Ford Delivery Sedan   

              productVendor  quantityOrdered priceEach  
0   Motor City Art Classics               50     55.09  
1   Motor City Art Classics               50    127.84  
2  Carousel DieCast Legends               50     75.47  
3       Second Gear Diecast               50     55.96  
4       Studio M Art Models               50     43.27  
5      Unimax Art Galleries               50     52.42  
6      Unimax Art Galleries               50    126.52  
7  Carousel DieCast Legends               50     43.27  
8   Motor City Art Classics               50    102.92  
9  Carousel DieCast Legends               50     76.31

- Below is an example of how a customer and their orders are stored in MySQL, and how the graph should look like in Neo4j
    - Note that the same order may be represented as many relationships since one order could contain many products
- The `Customer` nodes should have the following attributes:
    - `customerNumber`
    - `customerName`
    - `country`
- The `Product` nodes should have the following attributes:
    - `productCode`
    - `productName`
    - `productVendor`
- The `ORDERED` relationships should have the following attributes:
    - `orderNumber`
    - `orderDate`
    - `quantityOrdered`
    - `priceEach`


**MySQL relation:**

|    |   customerNumber | customerName              | country   |   orderNumber | orderDate   | productCode   | productName                             | productVendor             |   quantityOrdered |   priceEach |
|---:|-----------------:|:--------------------------|:----------|--------------:|:------------|:--------------|:----------------------------------------|:--------------------------|------------------:|------------:|
|  0 |              450 | The Sharp Gifts Warehouse | USA       |         10250 | 2004-05-11  | S32_4289      | 1928 Ford Phaeton Deluxe                | Highway 66 Mini Classics  |                50 |       62.6  |
|  1 |              450 | The Sharp Gifts Warehouse | USA       |         10257 | 2004-06-14  | S18_2949      | 1913 Ford Model T Speedster             | Carousel DieCast Legends  |                50 |       92.19 |
|  2 |              450 | The Sharp Gifts Warehouse | USA       |         10400 | 2005-04-01  | S10_4757      | 1972 Alfa Romeo GTA                     | Motor City Art Classics   |                64 |      134.64 |
|  3 |              450 | The Sharp Gifts Warehouse | USA       |         10400 | 2005-04-01  | S18_3856      | 1941 Chevrolet Special Deluxe Cabriolet | Exoto Designs             |                58 |       88.93 |
|  4 |              450 | The Sharp Gifts Warehouse | USA       |         10407 | 2005-04-22  | S18_1589      | 1965 Aston Martin DB5                   | Classic Metal Creations   |                59 |      114.48 |
|  5 |              450 | The Sharp Gifts Warehouse | USA       |         10407 | 2005-04-22  | S18_1749      | 1917 Grand Touring Sedan                | Welly Diecast Productions |                76 |      141.1  |
|  6 |              450 | The Sharp Gifts Warehouse | USA       |         10407 | 2005-04-22  | S18_4933      | 1957 Ford Thunderbird                   | Studio M Art Models       |                66 |       64.14 |
|  7 |              450 | The Sharp Gifts Warehouse | USA       |         10407 | 2005-04-22  | S24_1628      | 1966 Shelby Cobra 427 S/C               | Carousel DieCast Legends  |                64 |       45.78 |
|  8 |              450 | The Sharp Gifts Warehouse | USA       |         10407 | 2005-04-22  | S24_2766      | 1949 Jaguar XK 120                      | Classic Metal Creations   |                76 |       81.78 |
|  9 |              450 | The Sharp Gifts Warehouse | USA       |         10407 | 2005-04-22  | S24_2887      | 1952 Citroen-15CV                       | Exoto Designs             |                59 |       98.65 |


**Neo4j graph:**

<img src="./Neo4j-example.png">

In [18]:
# Deletes all customers and products (and their relationships). 
# Feel free to run this as many times as you want to reset your data.
_ = graph.execute_query("""
    match (c:Customer), (p:Product)
    detach delete c, p
""")

In [19]:
# TODO: Write and execute queries to create nodes and relationships

"""
Tips:

    To iterate through dataframe:
    
        for _, r in customer_orders_limit_df.iterrows():
            r = dict(r)
            Access fields like r['customerName'], r['country'], ...
            
    
    The priceEach field are stored as a Decimal object in the dataframe. This type is not 
    compatible with the neo4j API. You can convert it to a string by calling str(r['priceEach']). 


    You should call graph.execute_query to execute your queries. This method takes in a second
    optional argument, a dict. This allows you to do query parameters. For instance, to execute
    the query in the screenshot above, you could run
    
        graph.execute_query(
            "match (c:Customer { customerNumber: $custNum })-[:ORDERED]->(p:Product) return c, p",
            { "custNum": 450 }
        )

"""


for _, r in customer_orders_limit_df.iterrows():
    r = dict(r)
    graph.execute_query("""
        MERGE (c:Customer {customerNumber: $customerNumber})
        ON CREATE SET c.customerName = $customerName, c.country = $country
        MERGE (p:Product {productCode: $productCode})
        ON CREATE SET p.productName = $productName, p.productVendor = $productVendor
        MERGE (c)-[rel:ORDERED {
            orderNumber: $orderNumber,
            orderDate: $orderDate,
            quantityOrdered: $quantityOrdered,
            priceEach: $priceEachStr
        }]->(p)
    """, {
        "customerNumber": r['customerNumber'],
        "customerName": r['customerName'],
        "country": r['country'],
        "productCode": r['productCode'],
        "productName": r['productName'],
        "productVendor": r['productVendor'],
        "orderNumber": r['orderNumber'],
        "orderDate": str(r['orderDate']),  # Convert datetime to string if necessary
        "quantityOrdered": r['quantityOrdered'],
        "priceEachStr": str(r['priceEach'])  # Convert Decimal to string to ensure compatibility
    })

## Neo4j: Testing

- Run through the following cells
- **Make sure the outputs are fully visible**

In [20]:
res = []

for r in graph.execute_query("""
    match (c:Customer { customerNumber: 412 })-[o:ORDERED]->(p:Product)
    return c, o, p
""").records:
    res.append(dict(r['c']) | dict(r['o']) | dict(r['p']))
    
pandas.DataFrame(res)

country  customerNumber                   customerName  orderNumber  \
0  New Zealand             412  Extreme Desk Decorations, Ltd        10418   
1  New Zealand             412  Extreme Desk Decorations, Ltd        10234   
2  New Zealand             412  Extreme Desk Decorations, Ltd        10268   
3  New Zealand             412  Extreme Desk Decorations, Ltd        10418   

   quantityOrdered   orderDate priceEach productCode  \
0               52  2005-05-16     64.41    S24_2360   
1               50  2004-03-30    146.65    S18_1662   
2               50  2004-07-12    124.59    S18_2325   
3               50  2005-05-16    100.01    S32_4485   

                   productName             productVendor  
0      1982 Ducati 900 Monster  Highway 66 Mini Classics  
1  1980s Black Hawk Helicopter         Red Start Diecast  
2    1932 Model A Ford J-Coupe     Autoart Studio Design  
3    1974 Ducati 350 Mk3 Desmo       Second Gear Diecast

In [21]:
res = []

for r in graph.execute_query("""
    match (c:Customer)-[o:ORDERED]->(p:Product { productCode: 'S12_2823' })
    return c, o, p
""").records:
    res.append(dict(r['c']) | dict(r['o']) | dict(r['p']))
    
pandas.DataFrame(res)

country  customerNumber           customerName  orderNumber  \
0   France             353     Reims Collectables        10121   
1  Austria             382  Salzburg Collectables        10341   
2       UK             201  UK Collectables, Ltd.        10403   

   quantityOrdered   orderDate priceEach productCode       productName  \
0               50  2003-05-07    126.52    S12_2823  2002 Suzuki XREO   
1               55  2004-11-24    120.50    S12_2823  2002 Suzuki XREO   
2               66  2005-04-08    122.00    S12_2823  2002 Suzuki XREO   

          productVendor  
0  Unimax Art Galleries  
1  Unimax Art Galleries  
2  Unimax Art Galleries

In [22]:
res = []

for r in graph.execute_query("""
    match (c:Customer)-[o:ORDERED { quantityOrdered: 60 }]->(p:Product)
    return c, o, p
""").records:
    res.append(dict(r['c']) | dict(r['o']) | dict(r['p']))
    
pandas.DataFrame(res)

country  customerNumber               customerName  orderNumber  \
0      Spain             141     Euro+ Shopping Channel        10412   
1        USA             362          Gifts4AllAges.com        10414   
2  Australia             282  Souveniers And Things Co.        10420   

   quantityOrdered   orderDate priceEach productCode  \
0               60  2005-05-03    157.49    S18_3232   
1               60  2005-05-06     72.58    S24_3151   
2               60  2005-05-29     60.26    S24_1046   

                        productName         productVendor  
0       1992 Ferrari 360 Spider red  Unimax Art Galleries  
1  1912 Ford Model T Delivery Wagon       Min Lin Diecast  
2        1970 Chevy Chevelle SS 454  Unimax Art Galleries